File: 208-implement-trie-prefix-tree/implement-trie-prefix-tree.js
class TrieNode{
    constructor(){
        this.children={};
        this.isEndOfWord=false;
    }
}
var Trie=function(){
    this.root=new TrieNode();
};
/** 
 * @param {string} word
 * @return {void}
 */
Trie.prototype.insert=function(word){
    let node=this.root;
    for (const char of word){
        if(!(char in node.children)){
            node.children[char]=new TrieNode();
        }
        node=node.children[char];
    }
    node.isEndOfWord=true;
};

/** 
 * @param {string} word
 * @return {boolean}
 */
Trie.prototype.search=function(word){
    let node=this.root;
    for (const char of word){
        if(!(char in node.children)){
            return false;
        }
        node=node.children[char];
    }
    return node.isEndOfWord;
};

/** 
 * @param {string} prefix
 * @return {boolean}
 */
Trie.prototype.startsWith=function(prefix){
    let node=this.root;
    for (const char of prefix){
        if(!(char in node.children)){
            return false;
        }
        node=node.children[char];
    }
    return true;
};
/** 
 * Your Trie object will be instantiated and called as such:
 * var obj = new Trie()
 * obj.insert(word)
 * var param_2 = obj.search(word)
 * var param_3 = obj.startsWith(prefix)
 */

File: 208-implement-trie-prefix-tree/implement-trie-prefix-tree.cs
using System.Collections.Generic;

public class TrieNode
{
    public Dictionary<char, TrieNode> Children { get; private set; }
    public bool IsEndOfWord { get; set; }

    public TrieNode()
    {
        Children = new Dictionary<char, TrieNode>();
        IsEndOfWord = false;
    }
}

public class Trie
{
    private readonly TrieNode root;

    public Trie()
    {
        root = new TrieNode();
    }
    
    public void Insert(string word)
    {
        TrieNode current = root;
        foreach (var ch in word)
        {
            if (!current.Children.ContainsKey(ch))
            {
                current.Children[ch] = new TrieNode();
            }
            current = current.Children[ch];
        }
        current.IsEndOfWord = true;
    }
    
    public bool Search(string word)
    {
        TrieNode current = root;
        foreach (var ch in word)
        {
            if (!current.Children.ContainsKey(ch))
            {
                return false;
            }
            current = current.Children[ch];
        }
        return current.IsEndOfWord;
    }
    
    public bool StartsWith(string prefix)
    {
        TrieNode current = root;
        foreach (var ch in prefix)
        {
            if (!current.Children.ContainsKey(ch))
            {
                return false;
            }
            current = current.Children[ch];
        }
        return true;
    }
}

// Usage example:
// Trie trie = new Trie();
// trie.Insert("apple");
// Console.WriteLine(trie.Search("apple"));   // Output: True
// Console.WriteLine(trie.Search("app"));     // Output: False
// Console.WriteLine(trie.StartsWith("app")); // Output: True
// trie.Insert("app");
// Console.WriteLine(trie.Search("app"));     // Output: True
